### ISSUE TITLE
Wrong diff is shown during interactive rebase

### ISSUE THREAD

**Issue by Sergey-Makarov** [magit/magit#5265]
# Tested on

- Mac OS 14.5
- emacs-version: 29.2
- magit-version
  - 20240831.2255
  - 20241102.1300
  - 20241206.2214
- git version (git --version)
  - 2.47.1
  - 2.39.3 (Apple Git-146)

# Reproduction steps

Imagine we have a repo with the following commits (just as an example):

    5 File 2 change 2 <-- HEAD, main
    4 File 2 change 1
    3 File 2 created
    2 File 1 created
    1 Empty commit

I want to squash commits 4 and 5 to have the following structure:

    3_new File 2 created + changed
    2  File 1 created
    1  Empty commit

I do:

1.  Call magit-log
2.  Step onto commit 3
3.  Call magit-rebase -\> interactively
4.  pick commit 3, squash commits 4 and 5, and hit C-c C-c

Expected behaviour: Along with commit message buffer, I see a diff buffer showing a diff between 2 and what_is_supposed_to_be_3_new commits. (Changes from HEAD~1 to index)

Actual behaviour: Along with commit message buffer, I see a diff buffer showing a diff between **1** and what_is_supposed_to_be_3_new commits. (Changes from HEAD~2 to index)
In other words, the diff buffer also shows the creation of file 1, which is irrelevant. Although, from commit message buffer, it's clear that the ongoing commit only affects file 2:

      ..........
      ..........
    # Changes to be committed:
    # new file:   file2.txt
    #

# Side notes

1.  In Magit 3.3, only staged changes were shown (i.e. diff of commit 5) in a diff buffer, which makes me believe that in Magit 4+ the intention was to add relevant context into the diff buffer.
2.  If we hit C-c C-k when writing the commit message and then go magit-rebase -\> continue, we will only see staged changes in a diff buffer. I'd expect to see the same diff as during the steps above, but not 100% sure whether it should be the right behaviour :)

**Comment by tarsius** [comment]
Both of these diffs are potentially useful. The one you don't want shows all the changes that will be in the commit you are about to create. You want to only see the changes of n-1 commits.

While committing, type `C-c C-d` in the diff or commit message edit buffer to switch between these two diffs.

**Comment by Sergey-Makarov** [comment]
@tarsius Thanks for `C-c C-d` tip, did not know about that!

However, just to clarify, when I do `C-c C-d` twice (after doing all steps I described above), the diff goes from HEAD~2..index to essentially HEAD~1..index. Isn't it a wrong behaviour?

**Comment by Sergey-Makarov** [comment]
UPD: While testing [this](https://github.com/magit/magit/commit/7adad8c8d3bd61ae36659638751223cfa2c7d720) commit, I ran into another thing that seems weird:
If I launch Emacs and do the steps from my original message for the first time, `C-c C-d` cycles through `HEAD~2..index` and `HEAD^..index` diffs.

However, if I abort the rebase entirely and then do exactly the same steps, I see `HEAD~2..index` diff at first, and then `C-c C-d` cycles through `Staged changes` and `HEAD^..index`.

**Comment by tarsius** [comment]
I am aware that it doesn't work perfectly and I have tried a bunch of times to improve it. There is no way to ask Git what is going on and so we have to rely on heuristics to figure out what the current situation is.

You can look at `magit-commit-diff-1` if you want to take a stab at it yourself. I currently don't feel like giving it yet another try myself. I consider the current heuristics good enough, aka there are many more important things to deal with first. But it's already in my private "things to finish at some point" list.

**Comment by tarsius** [comment]
I had a quick look and noticed some things I *might* be able to improve somewhat quickly. (Some time next year.)

**Comment by Sergey-Makarov** [comment]
Great! In the meantime, I patched `magit-commit-diff-1` function locally like so:
``` lisp
  (defun magit-commit-diff-1 ()
    ...
      (pcase (list staged unstaged command)
        ...
        ;; ((and `(,_ ,_ magit-commit--rebase)
        ;;       (guard (integerp squash)))
        ;;  (setq rev (format "HEAD~%s" squash)))
        ...

```
Makes the job done for me 🙂 I don't see how this case might be useful (for me it's the opposite, plus it might go out of bounds of the commit history, especially if you squash a lot of commits), but maybe it covers a case I am not aware of 🙂

**Comment by pkryger** [comment]
FWIW I am observing the opposite to what has been mentioned in the opening post (with magit-20250808.949). That is, the diff for squashed range is missing a commit. I think that patch described in [comment above](https://github.com/magit/magit/issues/5265#issuecomment-2525240073) may not apply at this point in time, since `(magit-repository-local-get 'this-commit-command)` yields either `with-editor-finish` or `magit-rebase-continue`.
